server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -750,43 +748,70 @@ protected void doRun() {
-                                    } else {
-                                        // We were given a failure result in the onResponse method, so we must store the failure
-                                        // Recover the original document state, track a failed ingest, and pass it along
-                                        updateIndexRequestMetadata(indexRequest, originalDocumentMetadata);
-                                        totalMetrics.ingestFailed();
-                                        onStoreFailure.apply(slot, result.failedIndex, result.exception);
+                        final ActionListener<Boolean> documentListener = ActionListener.runAfter(new ActionListener<>() {
+                            @Override
+                            public void onResponse(Boolean shouldKeep) {
+                                assert shouldKeep != null;
+                                if (shouldKeep == false) {
+                                    String originalIndex = indexRequest.index();
+                                    // We only increment the total counter for dropped docs here, because these docs don't reach the code
+                                    // that ordinarily take care of that.
+                                    // We reuse `shouldStoreFailure` here to determine whether the index request targets a data stream,
+                                    // because we only want to track these metrics for data streams.
+                                    Optional<Boolean> storeFailure = shouldStoreFailure.apply(originalIndex);
+                                    if (storeFailure.isPresent()) {
+                                        failureStoreMetrics.incrementTotal(originalIndex);
                                     }
+                                    onDropped.accept(slot);
+                                } else {
+                                    assert firstPipeline != null;
+                                    firstPipeline.getMetrics().postIngestBytes(indexRequest.ramBytesUsed());
                                 }
+                            }
 
-                                @Override
-                                public void onFailure(Exception e) {
+                            @Override
+                            public void onFailure(Exception e) {
+                                final String originalIndex = indexRequest.index();
+                                String errorType = ElasticsearchException.getExceptionName(ExceptionsHelper.unwrapCause(e));
+                                // If `storeFailure` is true, we store the failure. If it's false, the target is a data stream,
+                                // but it doesn't have the failure store enabled. If it's empty, the target wasn't a data stream.
+                                Optional<Boolean> storeFailure = shouldStoreFailure.apply(originalIndex);
+                                if (storeFailure.orElse(false)) {
```

Comment:
Not really a fan of this `Optional<Boolean>` myself tbh. My struggle was to avoid having to check a second time whether the request targets a data stream when `storeFailure` is `false`.

Reply:
Is it that big of an issue?

Reply:
Hmm well we'd be doing it twice for every index request that doesn't have the failure store enabled. I'm usually inclined to pay extra care to doing things for every index request, but that's not based on any experimentation on the impact of doing a second lookup.

Alternatively, I could change `shouldStoreFailure` to a function that returns a failure store _candidate_ just like I do in `BulkOperation`. I'm also trying to think about the future where the decision to redirect a failure or not might be influenced by the request itself.